### PR TITLE
Knex auto-instrumentation 

### PR DESCRIPTION
**Pull request by owais**
This builds on top of: https://github.com/signalfx/signalfx-nodejs-tracing/pull/45

- Added support for both nested and flat promises when using knex.
- Added support for 0.18 series
- Re-worked context propagation to make auto-instrumentation work with all version >0.10
- Added test cases to cover edge cases I discovered or suspected would happen with the new context prop.

### CODE REVIEW

File: docs/API.md [hunk 1]
```diff
@@ -604,6 +604,17 @@ query HelloWorld {
...
+#### Knex
+
+##### Tags
+
+| Tag              | Description                                                |
+|------------------|------------------------------------------------------------|
+| component        | `knex`                                                     |
+| db.type          | The database driver being used such as sqlite3, etc        |
+| db.user          | The database user used by the driver to connect to the db. |
+| db.statement     | The SQL statement executed by the traced query.            |
```
Comment (rmfitzpatrick) [inline docs/API.md]: Would you recommend setting `db.instance` if reliably obtained?

File: src/plugins/knex.js [hunk 2]
```diff
@@ -2,22 +2,145 @@
...
-      this.unwrap(Builder.prototype, 'then')
+const spanSymbol = '_sfxSpan'
+
+function createWrapBuilder (tracer, config) {
+  return function wrapQueryBuilder (original) {
+    return function queryBuilderWithTrace () {
+      const scope = tracer.scope()
+      const span = scope.active()
+      const builder = original.apply(this, arguments)
+      builder[spanSymbol] = span
```
Comment (rmfitzpatrick) [inline src/plugins/knex.js]: nit: I believe the approach when modifying elements exposed to the user is to use `Object.defineProperty()` to keep them relatively closed.

File: src/plugins/knex.js [hunk 3]
```diff
@@ -2,22 +2,145 @@
...
+
+function createWrapRunnerQuery (tracer, config) {
+  return function wrapQuery (original) {
+    return function queryWithTrace (q) {
+      const scope = tracer.scope()
+      const childOf = this.builder[spanSymbol]
+
+      const tags = {
+        'component': 'knex',
+        'db.statement': q.sql
```
Comment (rmfitzpatrick) [inline src/plugins/knex.js]: Mind truncating to 1024 chars to prevent size issues?

File: src/plugins/knex.js [hunk 4]
```diff
@@ -2,22 +2,145 @@
...
+      },
+      unpatch (Client) {
+        this.unwrap(Client.prototype, 'runner')
+        this.unwrap(Client.prototype, 'raw')
+        this.unwrap(Client.prototype, 'schemaBuilder')
+        this.unwrap(Client.prototype, 'queryBuilder')
+      }
+    },
+    // Not sure if patching the following `then` methods help in anyway now but I'm leaving them
+    // in as they did solve some problems in the past and they don't seem to have any adverse effects
```
Comment (rmfitzpatrick) [inline src/plugins/knex.js]: I believe the intention is to ensure scope propagation where bluebird and native promise instrumentation fail.  What about your changes make this unnecessary?

File: docs/API.md (hunk 1 above)
Comment (owais) [reply docs/API.md]: Will look.

File: src/plugins/knex.js (hunk 2 above)
Comment (owais) [reply src/plugins/knex.js]: Thanks. Will do that.

File: src/plugins/knex.js (hunk 3 above)
Comment (owais) [reply src/plugins/knex.js]: I wouldn't. This doesn't include any "data" unless someone manually compiles queries and then uses them in an raw operation. Historically we haven't tried to enforce any sizing restrictions on tags at least, not on the client side. Happy to hear other thoughts.

File: src/plugins/knex.js (hunk 4 above)
Comment (owais) [reply src/plugins/knex.js]: My changes rely on the first call to a query builder to be able to extract context from the trace scope. Post that it passes the context on the builder itself. I removed these patches and wasn't didn't see any regressions but wasn't confident enough that they wouldn't be needed.

Comment (owais) [reply src/plugins/knex.js]: Ah, since only query builders needs the initial context from async scope, I can only patch that and get rid of everything else.

Comment (owais) [reply src/plugins/knex.js]: On second thought, I'm removing these as I'm unable to prove that they are actually needed. I couldn't break any tests or example apps without these patches. We can always add them back if we find a bug later.

File: docs/API.md (hunk 1 above)
Comment (owais) [reply docs/API.md]: Done

File: src/plugins/knex.js (hunk 3 above)
Comment (rmfitzpatrick) [reply src/plugins/knex.js]: The SFx offerings all (should) truncate db.statements to 1024 chars for its current [size restrictions](https://docs.signalfx.com/en/latest/apm/apm-metadata/apm-metadata.html#limits-on-span-metadata).  When we don't users have run into difficulties that are inconvenient to track down.

File: src/plugins/knex.js (hunk 2 above)
Comment (owais) [reply src/plugins/knex.js]: Done

File: src/plugins/knex.js (hunk 3 above)
Comment (owais) [reply src/plugins/knex.js]: Done

File: test/plugins/knex.spec.js [hunk 5]
```diff
@@ -5,6 +5,12 @@
 const agent = require('./agent')
 const plugin = require('../../src/plugins/knex')
 
+const sort = spans => spans.sort((a, b) => a.start.toString() >= b.start.toString() ? 1 : -1)
+
+function normalizeDBStatement (statement) {
```
Comment (sm7x) [inline test/plugins/knex.spec.js]: Everything looks great to me. However, are the statements produced through the tests different from those in real world situations? Was just wondering if this normalize function should be in the instrumentation instead.

Comment (owais) [reply test/plugins/knex.spec.js]: Older versions of knex use double quotes to quote things in queries while the newer ones use the backtick. This normalizes everything to one format so tests don't have to be too verbose and have multiple checks for multiple versions.

Comment (owais) [reply test/plugins/knex.spec.js]: So it is very specific to the tests. We don't want to modify the query in instrumentation code.

**Comment by owais** [comment]
@sm7x @flands any more concerns or can I go ahead and merge?
